deepvariant/channels/homopolymer_indel_quality_channel.cc

```diff
@@ -0,0 +1,194 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "deepvariant/channels/homopolymer_indel_quality_channel.h"
+
+#include <cmath>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "deepvariant/channels/channel.h"
+#include "deepvariant/protos/deepvariant.pb.h"
+#include "third_party/nucleus/protos/reads.pb.h"
+
+namespace learning {
+namespace genomics {
+namespace deepvariant {
+
+HomopolymerInDelQualityChannel::HomopolymerInDelQualityChannel(
+    int width, const PileupImageOptions& options)
+    : Channel(width, options) {}
+
+// Explanation of what TP means in Ultima Genomics data:
+// Given read R of size N,
+// Cram will have a QUAL vector of size N, and a TP vector of size N.
+//
+// TP[i] refers to the direction and magniture of the error which is encoded by
+// QUAL[i]
+//
+// For example:
+// If read has sequence ...TAAAAAG...,
+// decided_homopolymer_size for the A homopolymer is len(AAAAA) = 5.
+// Let i be an index in the read in the range of this homopolymer: [start, end].
+// When TP[i] = 1, it means QUAL[i] will contain the PHRED score for the
+// homopolymer to be 5 + 1 = 6.
+// When TP[i] = -1, it means QUAL[i] will contain the PHRED score for the
+// homopolymer to be 5 - 1 = 4.
+// When TP[i] = 2, it means QUAL[i] will contain the PHRED score for the
+// homopolymer to be 5 + 2 = 7.
+std::vector<int8_t> HomopolymerInDelQualityChannel::GetTPValues(
+    const Read& read) {
+  std::vector<int8_t> int_tps(read.aligned_sequence().size());
+  if (!read.info().contains("tp")) {
+    // Return a vector of zeros if tp tag is not present.
+    return int_tps;
+  }
+  const auto& tps = read.info().at("tp").values();
+
+  if (tps.empty()) {
+    return int_tps;
+  }
+
+  for (int i = 0; i < tps.size(); i++) {
+    int_tps[i] = tps[i].int_value();
+  }
+  return int_tps;
+}
+
+std::vector<std::uint8_t> HomopolymerInDelQualityChannel::HomoPolymerWeighted(
+    const Read& read) {
+  // Generates a vector reflecting the number of repeats observed
+  std::vector<std::uint8_t> homopolymer_weighted(read.aligned_sequence().size(),
+                                                 1);
+  const std::string& seq = read.aligned_sequence();
+
+  if (seq.empty()) {
+    return homopolymer_weighted;
+  }
+
+  int i = 0;
+  while (i < seq.size()) {
+    int hmer_length = 1;
+    char current_base = seq[i];
+    int j = i + 1;
+
+    // Count consecutive identical bases
+    while (j < seq.size() && seq[j] == current_base) {
+      hmer_length++;
+      j++;
+    }
+
+    // Fill all positions in this homopolymer with its length
+    for (int k = i; k < j; k++) {
+      homopolymer_weighted[k] = hmer_length;
```

Comment:
@doron-st , I'm going to change this to:

```python
      homopolymer_weighted[k] = std::min(hmer_length, kMaxHomoPolymerWeighted);
```

I want to check with you that makes sense?